core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java

```diff
@@ -352,64 +378,85 @@ private void runInferencingSparqlQueries(SailRepositoryConnection shaclSailConne
 	 *
 	 * @throws SailException if the thread is interrupted while waiting to obtain the lock.
 	 */
-	long acquireExclusiveWriteLock(long stamp) {
-		if (lock.validate(stamp)) {
-			return stamp;
+	Lock acquireExclusiveWriteLock(Lock lock) {
+
+		if (lock != null && lock.isActive()) {
+			return lock;
 		}
 
+		assert lock == null;
+
 		if (threadHoldingWriteLock == Thread.currentThread()) {
 			throw new SailConflictException(
 					"Deadlock detected when a single thread uses multiple connections " +
 							"interleaved and one connection has modified the shapes without calling commit() " +
 							"while another connection also tries to modify the shapes!");
 		}
 
-		long newStamp = lock.writeLock();
-		threadHoldingWriteLock = Thread.currentThread();
-		return newStamp;
-	}
-
-	boolean holdsWriteLock(long stamp) {
-		return lock.validate(stamp);
+		try {
+			Lock writeLock = lockManager.getWriteLock();
+			threadHoldingWriteLock = Thread.currentThread();
+			return writeLock;
+		} catch (InterruptedException e) {
+			throw new RuntimeException(e);
+		}
 	}
 
 	/**
 	 * Releases the exclusive write lock.
 	 *
 	 * @return
 	 */
-	long releaseExclusiveWriteLock(long stamp) {
-		assert stamp != 0;
+	Lock releaseExclusiveWriteLock(Lock lock) {
+		assert lock != null;
 		threadHoldingWriteLock = null;
-		lock.unlockWrite(stamp);
-		return 0;
+		lock.release();
+		return null;
 	}
 
-	long acquireReadlock() {
+	Lock acquireReadlock() {
 		if (threadHoldingWriteLock == Thread.currentThread()) {
 			throw new SailConflictException(
 					"Deadlock detected when a single thread uses multiple connections " +
 							"interleaved and one connection has modified the shapes without calling commit() " +
 							"while another connection calls commit()!");
 		}
-		return lock.readLock();
+		try {
+			return lockManager.getReadLock();
+		} catch (InterruptedException e) {
+			throw new RuntimeException(e);
+		}
+
 	}
 
-	long releaseReadlock(long stamp) {
-		assert stamp != 0;
+	Lock releaseReadlock(Lock lock) {
+		assert lock != null;
 
-		lock.unlockRead(stamp);
-		return 0;
-	}
+		lock.release();
 
-	synchronized long convertToReadLock(long writeLockStamp) {
-		assert writeLockStamp != 0;
-		long l = lock.tryConvertToReadLock(writeLockStamp);
-		assert l != 0;
-		threadHoldingWriteLock = null;
-		return l;
+		return null;
 	}
 
+//	Lock convertToReadLock(Lock writeLockStamp) {
```

Comment:
The rdf4j lock manager doesn't support downgrading of locks, I might consider adding this in the future.....so I thought I would keep this code around for a bit longer.